File: 0606-construct-string-from-binary-tree/0606-construct-string-from-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    void preorder(TreeNode root, StringBuilder res){
        
        if(root == null)    return;
        
        boolean isLeaf = (root.left == null && root.right == null) ? true : false;
        
        res.append("(");
        
        res.append(root.val);
        
        if(!isLeaf){
            
            if(root.left != null)  preorder(root.left, res);
            
            else    res.append("()");
            
            preorder(root.right, res);            
            
        }
        
        res.append(")");
        
    }
    
    public String tree2str(TreeNode root) {
        
        if(root == null)    return new String("");
        
        StringBuilder res = new StringBuilder();
        
        preorder(root, res);
        
        return res.substring(1, res.length() - 1);        
    }
}

File: 0606-construct-string-from-binary-tree/0606-construct-string-from-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void preorder(TreeNode* root, string&res){
        
        if(!root)   return;
        
        bool isLeaf = (!root->left && !root->right) ? true : false;
        
        res.push_back('(');
        
        res.append(to_string(root->val));
        
        if(!isLeaf){
            
            if(root->left)  preorder(root->left, res);

            else    res.append("()");

            if(root->right)     preorder(root->right, res);   
            
        }
        
        res.push_back(')');
    }
    string tree2str(TreeNode* root) {
        
        if(!root)   return "";
        
        string res = "";
        
        preorder(root, res);
        
        return res.substr(1, res.size() - 2);        
    }
};